programs/leap-util/actions/chain.cpp

```diff
@@ -62,5 +72,47 @@ int chain_actions::run_subcommand_build() {
       std::cout << fc::json::to_pretty_string(chainbase::environment()) << std::endl;
    }
 
+   return 0;
+}
+
+int chain_actions::run_subcommand_sstate() {
+   bfs::path state_dir = "";
+
+   // default state dir, if none specified
+   if(opt->sstate_state_dir.empty()) {
+      auto root = fc::app_path();
+      auto default_data_dir = root / "eosio" / "nodeos" / "data" ;
+      state_dir  = default_data_dir / config::default_state_dir_name;
+   }
+   else {
+      // adjust if path relative
+      state_dir = opt->sstate_state_dir;
+      if(state_dir.is_relative()) {
+         state_dir = bfs::current_path() / state_dir;
+      }
+   }
+   
+   auto shared_mem_path = state_dir / "shared_memory.bin";
+
+   if(!bfs::exists(shared_mem_path)) {
+      std::cerr << "Unable to read database status: file not found: " << shared_mem_path << std::endl;
+      return -1;
+   }
+
+   char header[chainbase::header_size];
+   std::ifstream hs(shared_mem_path.generic_string(), std::ifstream::binary);
+   hs.read(header, chainbase::header_size);
+   if(hs.fail()) {
+      std::cerr << "Unable to read database status: file invalid or corrupt" << shared_mem_path <<  std::endl;
+      return -1;
+   }
+
+   chainbase::db_header* dbheader = reinterpret_cast<chainbase::db_header*>(header);
+   if(dbheader->dirty) {
+      std::cout << "Database dirty flag is set, shutdown was not clean" << std::endl;
+      return -1;
+   }
```

Comment:
Should check that dbheader's `header_id` is expected value, as it's possible the format will change in the future

Reply:
thank you, done